### PR TITLE
RGW: test for exact match if the test is not persistent

### PR DESCRIPTION
**Pull request by AliMasarweh**
RGW: test for exact match if the test is not persistent

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
* there is still one "exact_match=False" in: test_ps_s3_notification_push_amqp_on_master
* can we try running the persistent notification tests, at least the ones that do not execrsise retries, with  "exact_match=True"?

**Comment by AliMasarweh** [comment]
Passing teuthology run:
https://pulpito.ceph.com/alimasa-2024-04-28_08:21:50-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/

File: src/test/rgw/bucket_notification/test_bn.py [hunk 1]
```diff
@@ -1168,108 +1168,181 @@ def test_ps_s3_notification_errors_on_master():
...
-    # start amqp receivers
-    exchange = 'ex1'
-    task1, receiver1 = create_amqp_receiver_thread(exchange, topic_name1)
-    task2, receiver2 = create_amqp_receiver_thread(exchange, topic_name2)
-    task1.start()
-    task2.start()
+    receiver = {}
+    receiver2 = {}
+    host = get_ip()
+    if endpoint_type == 'http':
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1183]: is it possible to merge the if-else statements here with the ones in line 1215?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 2]
```diff
@@ -1168,108 +1168,181 @@ def test_ps_s3_notification_errors_on_master():
...
+        key = bucket.new_key(str(i))
+        objects_size[key.name] = object_size
+        thr = threading.Thread(target=set_contents_from_string, args=(key, content,))
         thr.start()
         client_threads.append(thr)
     [thr.join() for thr in client_threads]
 
     time_diff = time.time() - start_time
-    print('average time for creation + qmqp notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
+    print('average time for creation + http notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1296]:
```
 print('average time for creation +'+endpoint_type+' notification is: ' + str(time_diff*1000/number_of_objects) + ' 
```

File: src/test/rgw/bucket_notification/test_bn.py [hunk 3]
```diff
@@ -1168,108 +1168,181 @@ def test_ps_s3_notification_errors_on_master():
...
     time.sleep(5)
 
-    # check amqp receiver
+    # check receiver
     keys = list(bucket.list())
-    print('total number of objects: ' + str(len(keys)))
-    receiver1.verify_s3_events(keys, exact_match=True)
-    receiver2.verify_s3_events(keys, exact_match=True)
+    receiver.verify_s3_events(keys, exact_match=True, deletions=False, expected_sizes=objects_size, etags=etags)
+    if endpoint_type != 'http':
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1304]: why special handling for http?
* we can change the test to use one receiver in all cases
* or we can have a test with 2 receivers, and use 2 http servers for that listening on 2 different ports

Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1304]: http server doesn't have ack levels while **kafka** and **ampq** do have different ack levels,
do you want to delete the second receivers for **kafka** and **amp**q, or add another receiver for http?

Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1304]: yes. lets use one topic and receiver in this test.
we can add different ack level test later

**Comment by AliMasarweh** [comment]
> there are other tests that should be made endpoint agnostic, do you want to handle them in this PR, or in a diffeent PR?

I will do the rest in a different PR
> there is still one "exact_match=False" in: test_ps_s3_notification_push_amqp_on_master

I made them end_point agnostic because I am not able to test ampq, so if http and kafka work then ampq should work

File: src/test/rgw/bucket_notification/test_bn.py (hunk 3 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1304]: Sure!

**Comment by AliMasarweh** [comment]
> in line 3737, why do we use "False" while we use "True" for the other endpoints?

do you mean in https://github.com/ceph/ceph/pull/56943/files#diff-83b1d2148e5a0483255b9af5f707fec576625a16c04aca191578d555ca336443R3737
this is not my function:
```
def persistent_notification(endpoint_type, conn, account=None):
    """ test pushing persistent notification """
    zonegroup = get_config_zonegroup()
    # create bucket
    bucket_name = gen_bucket_name()
    bucket = conn.create_bucket(bucket_name)
    topic_name = bucket_name + TOPIC_SUFFIX
    receiver = {}
    host = get_ip()
    if endpoint_type == 'http':
        # create random port for the http server
        host = get_ip_http()
        port = random.randint(10000, 20000)
        # start an http server in a separate thread
        receiver = HTTPServerWithEvents((host, port))
        endpoint_address = 'http://'+host+':'+str(port)
        endpoint_args = 'push-endpoint='+endpoint_address+'&persistent=true'
        # the http server does not guarantee order, so duplicates are expected
        exact_match = False # <========== line 3737
    elif endpoint_type == 'amqp':
```

**Comment by AliMasarweh** [comment]
> in line 3737, why do we use "False" while we use "True" for the other endpoints?

because it is a persistent http endpoint
```
   def persistent_notification(endpoint_type, conn, account=None):
    """ test pushing persistent notification """
    zonegroup = get_config_zonegroup()
    # create bucket
    bucket_name = gen_bucket_name()
    bucket = conn.create_bucket(bucket_name)
    topic_name = bucket_name + TOPIC_SUFFIX
    receiver = {}
    host = get_ip()
    if endpoint_type == 'http':
        # create random port for the http server
        host = get_ip_http()
        port = random.randint(10000, 20000)
        # start an http server in a separate thread
        receiver = HTTPServerWithEvents((host, port))
        endpoint_address = 'http://'+host+':'+str(port)
        endpoint_args = 'push-endpoint='+endpoint_address+'&persistent=true'
        # the http server does not guarantee order, so duplicates are expected
        exact_match = False
    elif endpoint_type == 'amqp':
```

**Comment by yuvalif** [comment]
> > in line 3737, why do we use "False" while we use "True" for the other endpoints?
> 
> because it is a persistent http endpoint
> 
> ```
>    def persistent_notification(endpoint_type, conn, account=None):
>     """ test pushing persistent notification """
>     zonegroup = get_config_zonegroup()
>     # create bucket
>     bucket_name = gen_bucket_name()
>     bucket = conn.create_bucket(bucket_name)
>     topic_name = bucket_name + TOPIC_SUFFIX
>     receiver = {}
>     host = get_ip()
>     if endpoint_type == 'http':
>         # create random port for the http server
>         host = get_ip_http()
>         port = random.randint(10000, 20000)
>         # start an http server in a separate thread
>         receiver = HTTPServerWithEvents((host, port))
>         endpoint_address = 'http://'+host+':'+str(port)
>         endpoint_args = 'push-endpoint='+endpoint_address+'&persistent=true'
>         # the http server does not guarantee order, so duplicates are expected
>         exact_match = False
>     elif endpoint_type == 'amqp':
> ```

but the other endpoints there are also persistent, and they have `exact_match=True` ?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 4]
```diff
@@ -1168,108 +1168,181 @@ def test_ps_s3_notification_errors_on_master():
...
-    topic_conf2.del_config()
+    topic_conf.del_config()
+    if endpoint_type != 'http':
+        topic_conf2.del_config()
     # delete the bucket
     conn.delete_bucket(bucket_name)
+    if endpoint_type == 'http':
+        receiver.close()
+    else:
+        stop_amqp_receiver(receiver, task)
```
[the remark this answers is not in the document]
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1302]: How do we do that?
**stop_kafka_receiver** and **stop_ampq_receiver** are different
```
def stop_kafka_receiver(receiver, task):
    """stop the receiver thread and wait for it to finish"""
    receiver.stop = True
    task.join(1)
    try:
        receiver.consumer.unsubscribe()
        receiver.consumer.close()
    except Exception as error:
        log.info('failed to gracefully stop Kafka receiver: %s', str(error))
```
```
def stop_amqp_receiver(receiver, task):
    """stop the receiver thread and wait for it to finis"""
    try:
        receiver.channel.stop_consuming()
        log.info('stopping AMQP receiver')
    except Exception as error:
        log.info('failed to gracefully stop AMQP receiver: %s', str(error))
    task.join(5)
```

[the remark this answers is not in the document]
Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1302]: make it a member function of the receiver called "stop()" and have a different implementation per receiver type

**Comment by AliMasarweh** [comment]
teuthology run failing on a known issue:
http://qa-proxy.ceph.com/teuthology/alimasa-2024-05-05_14:42:26-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/7691982/teuthology.log

**Comment by AliMasarweh** [comment]
> > > in line 3737, why do we use "False" while we use "True" for the other endpoints?
> > 
> > 
> > because it is a persistent http endpoint
> > ```
> >    def persistent_notification(endpoint_type, conn, account=None):
> >     """ test pushing persistent notification """
> >     zonegroup = get_config_zonegroup()
> >     # create bucket
> >     bucket_name = gen_bucket_name()
> >     bucket = conn.create_bucket(bucket_name)
> >     topic_name = bucket_name + TOPIC_SUFFIX
> >     receiver = {}
> >     host = get_ip()
> >     if endpoint_type == 'http':
> >         # create random port for the http server
> >         host = get_ip_http()
> >         port = random.randint(10000, 20000)
> >         # start an http server in a separate thread
> >         receiver = HTTPServerWithEvents((host, port))
> >         endpoint_address = 'http://'+host+':'+str(port)
> >         endpoint_args = 'push-endpoint='+endpoint_address+'&persistent=true'
> >         # the http server does not guarantee order, so duplicates are expected
> >         exact_match = False
> >     elif endpoint_type == 'amqp':
> > ```
> 
> but the other endpoints there are also persistent, and they have `exact_match=True` ?

in the test it says that kafka and ampq guarantee ordering

```
    if endpoint_type == 'http':
        # create random port for the http server
        host = get_ip_http()
        port = random.randint(10000, 20000)
        # start an http server in a separate thread
        receiver = HTTPServerWithEvents((host, port))
        endpoint_address = 'http://'+host+':'+str(port)
        endpoint_args = 'push-endpoint='+endpoint_address+'&persistent=true'
        # the http server does not guarantee order, so duplicates are expected
        exact_match = False
    elif endpoint_type == 'amqp':
        # start amqp receiver
        exchange = 'ex1'
        task, receiver = create_amqp_receiver_thread(exchange, topic_name)
        task.start()
        endpoint_address = 'amqp://' + host
        endpoint_args = 'push-endpoint='+endpoint_address+'&amqp-exchange='+exchange+'&amqp-ack-level=broker'+'&persistent=true'
        # amqp broker guarantee ordering
        exact_match = True
    elif endpoint_type == 'kafka':
        # start kafka receiver
        task, receiver = create_kafka_receiver_thread(topic_name)
        task.start()
        endpoint_address = 'kafka://' + host
        endpoint_args = 'push-endpoint='+endpoint_address+'&kafka-ack-level=broker'+'&persistent=true'
        # kafka broker guarantee ordering
        exact_match = True
```

**Comment by yuvalif** [comment]
> in the test it says that kafka and ampq guarantee ordering

we never check the order of events. `exact_match=True` just mean we did not get duplicates

**Comment by yuvalif** [comment]
> teuthology run failing on a known issue: http://qa-proxy.ceph.com/teuthology/alimasa-2024-05-05_14:42:26-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/7691982/teuthology.log

is this the known crash?
can you please provide the crash backtrace?

**Comment by AliMasarweh** [comment]
> > teuthology run failing on a known issue: http://qa-proxy.ceph.com/teuthology/alimasa-2024-05-05_14:42:26-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/7691982/teuthology.log
> 
> is this the known crash? can you please provide the crash backtrace?

```
2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:======================================================================
2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:FAIL: test persistent topic stats
2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:----------------------------------------------------------------------
2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:Traceback (most recent call last):
2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    self.test(*self.arg)
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3044, in test_ps_s3_persistent_topic_stats
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    wait_for_queue_to_drain(topic_name)
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 2962, in wait_for_queue_to_drain
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    assert_equal(entries, 0)
2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:AssertionError: 200 != 0
```

**Comment by AliMasarweh** [comment]
> > in the test it says that kafka and ampq guarantee ordering
> 
> we never check the order of events. `exact_match=True` just mean we did not get duplicates

so we should make `exact_match=False`, right?

**Comment by yuvalif** [comment]
> > > teuthology run failing on a known issue: http://qa-proxy.ceph.com/teuthology/alimasa-2024-05-05_14:42:26-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/7691982/teuthology.log
> > 
> > 
> > is this the known crash? can you please provide the crash backtrace?
> 
> ```
> 2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:======================================================================
> 2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:FAIL: test persistent topic stats
> 2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:----------------------------------------------------------------------
> 2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:Traceback (most recent call last):
> 2024-05-05T15:42:12.838 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    self.test(*self.arg)
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 3044, in test_ps_s3_persistent_topic_stats
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    wait_for_queue_to_drain(topic_name)
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:  File "/home/ubuntu/cephtest/ceph/src/test/rgw/bucket_notification/test_bn.py", line 2962, in wait_for_queue_to_drain
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:    assert_equal(entries, 0)
> 2024-05-05T15:42:12.839 INFO:teuthology.orchestra.run.smithi150.stderr:AssertionError: 200 != 0
> ```

this is just the test failure. is it due to a rgw crash?

**Comment by yuvalif** [comment]
> > > in the test it says that kafka and ampq guarantee ordering
> > 
> > 
> > we never check the order of events. `exact_match=True` just mean we did not get duplicates
> 
> so we should make `exact_match=False`, right?

yes, lets be consistent with all endpoint types here

File: src/test/rgw/bucket_notification/test_bn.py [hunk 5]
```diff
@@ -1168,108 +1174,141 @@ def test_ps_s3_notification_errors_on_master():
...
+        key = bucket.new_key(str(i))
+        objects_size[key.name] = object_size
+        thr = threading.Thread(target=set_contents_from_string, args=(key, content,))
         thr.start()
         client_threads.append(thr)
     [thr.join() for thr in client_threads]
 
     time_diff = time.time() - start_time
-    print('average time for creation + qmqp notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
+    print('average time for creation + http notification is: ' + str(time_diff*1000/number_of_objects) + ' milliseconds')
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1271]: the log message should not say "http"

File: src/test/rgw/bucket_notification/test_bn.py [hunk 6]
```diff
@@ -1168,108 +1174,141 @@ def test_ps_s3_notification_errors_on_master():
...
+        s3_notification_conf = PSNotificationS3(conn, bucket_name, topic_conf_list)
+        response, status = s3_notification_conf.set_config()
+        assert_equal(status/100, 2)
+    else:
+        return SkipTest('Unknown endpoint type: ' + endpoint_type)
 
-    # create objects in the bucket (async)
-    number_of_objects = 100
+    # create objects in the bucket
+    number_of_objects = 10
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1253]: why change from 100 to 10?

File: src/test/rgw/bucket_notification/test_bn.py (hunk 5 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1271]: Ack

**Review by yuvalif — APPROVED** [review]

**Comment by AliMasarweh** [comment]
teuthology test:
https://pulpito.ceph.com/alimasa-2024-05-06_13:54:15-rgw:notifications-wip-alimasa-bn-test-distro-default-smithi/
failing on a known issue of https://tracker.ceph.com/issues/65337

**Comment by AliMasarweh** [comment]
jenkins test api

**Pull request merged**